cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -268,11 +272,24 @@ protected void processDag(PhaseSpec phaseSpec, String sourcePluginType, JavaSpar
         }
       }, Exception.class);
 
-      //We should have all the sink runnables at this point, execute them
-      executeSinkRunnables(sec, sinkRunnables);
+      // Exception is thrown from sink runnables to terminate the pipeline on a batch failure.
+      // But in most cases it was observed that the next pending batch starts processing while the shutdown is ongoing.
+      // Set the failure status in StreamingContext so that further batches do not save state even if they succeed.
+      // Without this, there is a risk of a batch being lost and at least once guarantee will fail.
+      try {
+        //We should have all the sink runnables at this point, execute them
+        executeSinkRunnables(sec, sinkRunnables);
+      } catch (Exception e) {
+        streamingContext.setFlag(BATCH_FAILED, true);
```

Comment:
is there a reason we need to do it through the streaming context instead of using that failedBatch boolean variable? 

If so it would be good to add a comment around why it is needed, and remove the failedBatch variable.

Reply:
changed to use an AtomicBoolean.